EventsByTag.md

```diff
@@ -0,0 +1,120 @@
+
+Details for each of the four combinations of current/live and off set/no offset:
+* First event missing: The first found does not have sequence number 1
+* Event missing: Find some events in order then a sequence number jump
+* Omitted events: Can events be omitted or will the stream fail
+
+
+#### Current query / No offset
+
+* First event missing: Back track to `first-time-bucket`. N scans of each bucket.
+* Event missing: N scans from the previous event for this persistenceId/tag's offset to the current offset.
+* Omitted events: No, stream will fail if either of the above two scenarios don't find the missing event.
+
+#### Current query / Offset
+
+* First event missing: Back track to Offset. N scans of each bucket then assume it is the first event from this offset.
+* Event missing: N scans from the previous event for this persistenceId/tag's offset to the current offset.
+* Omitted events: Yes. If first event missing doesn't find an event that should be there.
+
+
+#### Live query / No Offset
+
+* First event missing: Back track to `first-time-bucket`. N scans of each bucket.
+* Event missing: N scans from the previous event for this persistenceId/tag's offset to the current offset.
+* Omitted events: No, stream will fail if either of the above two scenarios don't find the missing event.
+
+#### Live query / Offset
+
+* First event missing: Back track to Offset. N scans of each bucket then assume it is the first event from this offset.
+* Event missing: N scans from the previous event for this persistenceId/tag's offset to the current offset.
+* Omitted events: Yes. If first event missing doesn't find an event that should be there.
+
+## Implementation
+
+A `TagWriter` actor is created, on demand, per tag that batches (unlogged batch grouped by the partition key)
+up writes to a separate table partitioned by tag name and a time bucket.
+
+When a message is persisted, if it has tags, it is sent to each of the `TagWriter`s asynchronously.
+
+This allows us to be more efficient and thus support more tags than the old version (limited to 3). No limit is 
+imposed with new implementation but it is advised to keep it small i.e. less than 10.
+
+
+### Crash recovery
+
+`TagWriter` maintains a separate table with how far it has gotten up to
+for each persistenceId.
+
+```
+CREATE TABLE akka.tag_views_progress (                              
+    peristence_id text,           
+    tag text,                     
+    sequence_nr bigint,           
+    PRIMARY KEY (peristence_id, tag)                                
+)
+```
+
+Part of recovery will then check that all events have been written by the tag writer.
+
+On the happy path this table will be written to after the tag_view write is complete
+so it could be lost in a crash. In that event recovery will duplicate writes to the 
+tag_views table which is fine as they will be upserts.
+
+TODO:
+What happens if a persistenceId is not uses for a long time after a crash? eventsByTag queries 
```

Comment:
Would it make sense to combine this with [cluster sharding remembering](https://doc.akka.io/docs/akka/2.5/scala/cluster-sharding.html#cluster-sharding-remembering), so the in-progress persistent actors would come back after cluster restoration? 

Actually, come to think of it, just using cluster sharding + persistent actor would perhaps _already_ work around this, as a "crash" would typically be just the one node, and hence another would take over that shard, correct?

Reply:
Yes it would be fixed by using remembering but have to worry about users that aren't using it